elide-graphql/src/main/java/com/yahoo/elide/graphql/GraphQLEntityProjectionMaker.java

```diff
@@ -0,0 +1,457 @@
+     * (See {@link #withParentEntity(Class, Supplier)}).
+     */
+    @Getter(AccessLevel.PRIVATE)
+    private final Deque<Class<?>> parentEntityStack = new LinkedList<>();
+
+    /**
+     * Reference
+     */
+    @Getter(AccessLevel.PRIVATE)
+    @Setter(AccessLevel.PRIVATE)
+    private Object valueWithVariable;
+
+    @NonNull
+    @Getter(AccessLevel.PRIVATE)
+    private final EntityDictionary dictionary;
+
+    @NonNull
+    @Getter(AccessLevel.PRIVATE)
+    private final ElideSettings elideSettings;
+
+    public Collection<EntityProjection> make(String query) {
+        ANTLRInputStream inputStream = new ANTLRInputStream(query);
+        GraphqlLexer lexer = new GraphqlLexer(inputStream);
+        CommonTokenStream tokens = new CommonTokenStream(lexer);
+        GraphqlParser parser = new GraphqlParser(tokens);
+
+        // GraphQL grammar on document (6.0):
+        //
+        //     document : definition+;
+        //
+        // Given that GraphQL can query multiple entities in a single query,
+        // we define one root level entity projection for each document.definition
+        List<GraphqlParser.DefinitionContext> definitions = parser.document().definition();
+
+        definitions.forEach(this::visitDefinition);
+
+        return allEntityProjections;
+    }
+
+    @Override
+    public Void visitDefinition(final GraphqlParser.DefinitionContext ctx) {
+        // GraphQL grammar on a single definition (6.0):
+        //
+        // definition:
+        //     operationDefinition |
+        //     fragmentDefinition |
+        //     typeSystemDefinition
+        //     ;
+
+        if (ctx.typeSystemDefinition() != null) {
+            String message = "Type System Definition is not allowed in read-only operations";
+            log.error(message);
+            throw new IllegalStateException(message);
+        }
+
+        // nothing interesting about entity projection here; go to child nodes down the parse tree
+        return super.visitDefinition(ctx);
+    }
+
+    @Override
+    public Void visitField(final GraphqlParser.FieldContext ctx) {
+        // GraphQL grammar for field (6.0):
+        //     field : alias? name arguments? directives? selectionSet?;
+
+        String nodeName = ctx.name().getText();
+        if (ConnectionContainer.EDGES_KEYWORD.equals(nodeName)
+                || EdgesContainer.NODE_KEYWORD.equals(nodeName)
+                || ConnectionContainer.PAGE_INFO_KEYWORD.equals(nodeName)
+        ) {
+            // do nothing on "edges", "node", or "pageInfo" field 
+            // see https://graphql.org/learn/pagination/#pagination-and-edges
+            // https://elide.io/pages/guide/11-graphql.html#relationship-metadata
+            return super.visitField(ctx);
+        }
+
+        // on relationship field, change "authors" to "author" so we can look it up in entity dictionary
+        String entityName = singularized(nodeName);
```

Comment:
I don't think this it is safe to assume any modifications to a relationship name.  You can look up the type by a combination of the parent class and the relationship name.